### PR TITLE
BUG:  Address N4 mask issue.

### PR DESCRIPTION
**Pull request by ntustison**
I think this should do it.

### CODE REVIEW

**Comment by ntustison** [comment]
@stnava , @muschellij2 --- it looks like the failures stem from [this ANTs pull request](https://github.com/ANTsX/ANTs/commit/76b0cc051bc0edefc9c638969264d23f36bdce6c) due to a switch in ITK from using ``Type`` to ``Enum``.  I'm assuming this will have to be updated in ITKR.  Do you want me to do that?

**Comment by muschellij2** [comment]
I would appreciate it as I would not know how to best address this.  I
don't know if @stnava has any concerns of breaks?
Best,
John

**Comment by ntustison** [comment]
Thanks @muschellij2 .  I don't have much familiarity with pull requests for ITKR so I'll wait until @stnava weighs in.

**Comment by muschellij2** [comment]
Got it - I misunderstood.  Do you want me to change the ITK tag in ITKR
to bcd762e7bd743341c2dce84852ced49d8292f293 ?
Best,
John

**Comment by muschellij2** [comment]
Also you need to re-document as the NAMESPACE was deleted.

**Comment by ntustison** [comment]
@muschellij2 Sorry about that---yeah, if you could clean up the mess I created, that would be great.

**Comment by ntustison** [comment]
Hey @muschellij2 and @stnava --- getting back to the N4 segfault issue.  @stnava merged the [updated ITKR](https://github.com/stnava/ITKR/commit/53d4f3d428cb827805486c882da0678136d82de4) with the new ITK git tag.  You can see that the travis builds are still failing due to the same issue.  Is there someplace in ANTsRCore where I need to update the ITKR git tag?

File: NAMESPACE [hunk 1]
```diff
@@ -1,224 +1,5 @@
 # Generated by roxygen2: do not edit by hand
 
-S3method(apply,antsImage)
```
[the remark this answers is not in the document]
Comment (ntustison) [reply NAMESPACE]: Sure.  I have no idea why these lines were deleted or even if this pull is still needed.

**Comment by muschellij2** [comment]
I don't think it is needed, we can close
Best,
John

**Comment by muschellij2** [comment]
This still segfaults for me after fresh install yesterday.   I can't do a `reprex` because it segfaults:

```r
sessioninfo::session_info("ANTsRCore")
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.2 (2019-12-12)
 os       macOS Mojave 10.14.6        
 system   x86_64, darwin15.6.0        
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/New_York            
 date     2020-03-19                  

─ Packages ───────────────────────────────────────────────────────────────────
 package   * version   date       lib source                          
 ANTsRCore   0.7.3.1   2020-03-19 [1] Github (ANTsX/ANTsRCore@5749f7c)
 ITKR        0.5.3     2020-03-18 [1] Github (stnava/ITKR@715331d)    
 lattice     0.20-38   2018-11-04 [1] CRAN (R 3.6.2)                  
 magrittr    1.5       2014-11-22 [1] CRAN (R 3.6.0)                  
 Matrix      1.2-18    2019-11-27 [1] CRAN (R 3.6.2)                  
 Rcpp        1.0.4.2   2020-03-19 [1] Github (RcppCore/Rcpp@20e462f)  
 RcppEigen   0.3.3.7.0 2019-11-16 [1] CRAN (R 3.6.0)                  

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```


https://travis-ci.org/github/ANTsX/ANTsRCore/jobs/662685464

**Comment by ntustison** [comment]
Is this still the example that faults?

```
fname = getANTsRData("r16")
in_img = antsImageRead(fname)
full_image = in_img > -1
n4 = n4BiasFieldCorrection(in_img, mask = full_image)
```

If it does, I just updated and ran it on two of my macs and our dgx.  I can't reproduce.   

On the DGX:
```
> fname = getANTsRData("r16")
> in_img = antsImageRead(fname)
> full_image = in_img > -1
> n4 = n4BiasFieldCorrection(in_img, mask = full_image)
> sessioninfo::session_info( "ANTsRCore" )
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.1 (2019-07-05)
 os       Ubuntu 18.04.2 LTS          
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/New_York            
 date     2020-03-19                  

─ Packages ───────────────────────────────────────────────────────────────────
 package   * version   date       lib source                          
 ANTsRCore * 0.7.3.1   2020-03-19 [1] Github (ANTsX/ANTsRCore@5749f7c)
 ITKR        0.5.3     2020-03-19 [1] Github (stnava/ITKR@715331d)    
 lattice     0.20-38   2018-11-04 [4] CRAN (R 3.6.1)                  
 magrittr    1.5       2014-11-22 [2] CRAN (R 3.6.1)                  
 Matrix      1.2-17    2019-03-22 [4] CRAN (R 3.6.1)                  
 Rcpp        1.0.4     2020-03-17 [1] CRAN (R 3.6.1)                  
 RcppEigen   0.3.3.7.0 2019-11-16 [1] CRAN (R 3.6.1)                  

[1] /home/ntustison/R/x86_64-pc-linux-gnu-library/3.6
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library
```

On my newish iMac:
```
> fname = getANTsRData("r16")
> in_img = antsImageRead(fname)
> full_image = in_img > -1
> n4 = n4BiasFieldCorrection(in_img, mask = full_image)
> sessioninfo::session_info( "ANTsRCore" )
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.2 (2019-12-12)
 os       macOS Catalina 10.15.3      
 system   x86_64, darwin15.6.0        
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/Los_Angeles         
 date     2020-03-19                  

─ Packages ───────────────────────────────────────────────────────────────────
 package   * version   date       lib source                          
 ANTsRCore * 0.7.3.1   2020-03-19 [1] Github (ANTsX/ANTsRCore@5749f7c)
 ITKR        0.5.3     2020-03-19 [1] Github (stnava/ITKR@715331d)    
 lattice     0.20-38   2018-11-04 [1] CRAN (R 3.6.2)                  
 magrittr    1.5       2014-11-22 [1] CRAN (R 3.6.0)                  
 Matrix      1.2-18    2019-11-27 [1] CRAN (R 3.6.2)                  
 Rcpp        1.0.4.3   2020-03-19 [1] Github (RcppCore/Rcpp@5ea44f1)  
 RcppEigen   0.3.3.7.0 2019-11-16 [1] CRAN (R 3.6.0)                  

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

Perhaps @stnava can see if it fails for him

**Comment by ntustison** [comment]
Since you're running Mojave, I'll also include my older iMac (running High Sierra):

```
> fname = getANTsRData("r16")
> in_img = antsImageRead(fname)
> full_image = in_img > -1
> n4 = n4BiasFieldCorrection(in_img, mask = full_image)
> sessioninfo::session_info( "ANTsRCore" )
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.3 (2020-02-29)
 os       macOS High Sierra 10.13.6   
 system   x86_64, darwin15.6.0        
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/Los_Angeles         
 date     2020-03-19                  

─ Packages ───────────────────────────────────────────────────────────────────
 package   * version   date       lib source                          
 ANTsRCore * 0.7.3.1   2020-03-19 [1] Github (ANTsX/ANTsRCore@5749f7c)
 ITKR        0.5.3     2020-03-19 [1] Github (stnava/ITKR@715331d)    
 lattice     0.20-38   2018-11-04 [1] CRAN (R 3.6.3)                  
 magrittr    1.5       2014-11-22 [1] CRAN (R 3.6.0)                  
 Matrix      1.2-18    2019-11-27 [1] CRAN (R 3.6.3)                  
 Rcpp        1.0.4.3   2020-03-19 [1] Github (RcppCore/Rcpp@5ea44f1)  
 RcppEigen   0.3.3.7.0 2019-11-16 [1] CRAN (R 3.6.0)                  

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

**Comment by muschellij2** [comment]
That works fine for me.

This one fails (only difference is the mask > 0 vs. > -1):
fname = getANTsRData("r16")
in_img = antsImageRead(fname)
mask = in_img > 0
n4 = n4BiasFieldCorrection(in_img, mask = mask)

Best,
John

**Comment by ntustison** [comment]
Thanks @muschellij2 .  Unfortunately, all three of my machines also work with ``in_img > 0``.

**Comment by muschellij2** [comment]
Still failing on mine, just reinstalled ANTsRCore.  I don't really know
what to do here, but it's making the n4 bias field correction unusable.

> library(ANTsRCore)
fname = getANTsRData("r16")

in_img = antsImageRead(fname)

full_image = in_img > -1

n4 = n4BiasFieldCorrection(in_img, mask = full_image)


Attaching package: ‘ANTsRCore’

The following objects are masked from ‘package:stats’:

    sd, var

The following objects are masked from ‘package:base’:

    all, any, apply, max, min, prod, range, sum

> fname = getANTsRData("r16")
>
> in_img = antsImageRead(fname)
>
> full_image = in_img > -1
>


THIS WORKS FINE

> n4 = n4BiasFieldCorrection(in_img, mask = full_image)
>

SEGFAULTS
> n4 = n4BiasFieldCorrection(in_img, mask = in_img > 0)

 *** caught segfault ***

 *** caught segfault ***
address 0x10d006000, cause 'memory not mapped'

 *** caught segfault ***
address 0x10d012000, cause 'memory not mapped'
address 0x10d00a000, cause 'memory not mapped'

Traceback:

Traceback:
 1:  1: .helpn4BiasFieldCorrection(args).helpn4BiasFieldCorrection(args)


 *** caught segfault ***
 2:  2: address 0x10d00e000, cause 'memory not mapped'
Traceback:
n4BiasFieldCorrection(in_img, mask = in_img > 0)
n4BiasFieldCorrection(in_img, mask = in_img > 0) 1:
.helpn4BiasFieldCorrection(args)


Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace




I am on 3.6.2, so maybe that could be a reason?  I highly doubt it:

> sessioninfo::session_info("ANTsRCore")
─ Session info
───────────────────────────────────────────────────────────────
 setting  value
 version  R version 3.6.2 (2019-12-12)
 os       macOS Mojave 10.14.6
 system   x86_64, darwin15.6.0
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2020-03-20

─ Packages
───────────────────────────────────────────────────────────────────
 package   * version   date       lib source
 ANTsRCore   0.7.3.1   2020-03-20 [1] Github (muschellij2/ANTsRCore@b6218e1)
 ITKR        0.5.3     2020-03-18 [1] Github (stnava/ITKR@715331d)
 lattice     0.20-38   2018-11-04 [1] CRAN (R 3.6.2)
 magrittr    1.5       2014-11-22 [1] CRAN (R 3.6.0)
 Matrix      1.2-18    2019-11-27 [1] CRAN (R 3.6.2)
 Rcpp        1.0.4.3   2020-03-20 [1] Github (RCppCore/Rcpp@5ea44f1)
 RcppEigen   0.3.3.7.0 2019-11-16 [1] CRAN (R 3.6.0)

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library

**Comment by ntustison** [comment]
Can you set ``verbose = TRUE`` and copy-and-paste the output?

**Comment by muschellij2** [comment]
``` r
library(ANTsRCore)
#> 
#> Attaching package: 'ANTsRCore'
#> The following objects are masked from 'package:stats':
#> 
#>     sd, var
#> The following objects are masked from 'package:base':
#> 
#>     all, any, apply, max, min, prod, range, sum
fname = getANTsRData("r16")
in_img = antsImageRead(fname)
n4 = n4BiasFieldCorrection(in_img)
rm(n4); gc()
#>           used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells 1849075 98.8    3536740 188.9         NA  2687578 143.6
#> Vcells 3302506 25.2    8388608  64.0      16384  5652218  43.2
mask = in_img > -1
mask
#> antsImage
#>   Pixel Type          : unsigned char 
#>   Components Per Pixel: 1 
#>   Dimensions          : 256x256 
#>   Voxel Spacing       : 1x1 
#>   Origin              : 0 0 
#>   Direction           : 1 0 0 1
sum(mask)
#> [1] 65536
n4 = n4BiasFieldCorrection(in_img, mask = mask, verbose = TRUE)
```

<sup>Created on 2020-03-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

<details style="margin-bottom:10px;">

<summary>Session info</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.3 (2020-02-29)
#>  os       macOS Mojave 10.14.6        
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2020-03-20                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date       lib source                             
#>  ANTsRCore   * 0.7.3.1     2020-03-20 [1] Github (ANTsX/ANTsRCore@c53aaaa)   
#>  assertthat    0.2.1       2019-03-21 [1] CRAN (R 3.6.0)                     
#>  backports     1.1.5       2019-10-02 [1] CRAN (R 3.6.0)                     
#>  cli           2.0.2       2020-02-28 [1] CRAN (R 3.6.0)                     
#>  crayon        1.3.4       2017-09-16 [1] CRAN (R 3.6.0)                     
#>  digest        0.6.25      2020-02-23 [1] CRAN (R 3.6.0)                     
#>  evaluate      0.14        2019-05-28 [1] CRAN (R 3.6.0)                     
#>  fansi         0.4.1       2020-01-08 [1] CRAN (R 3.6.0)                     
#>  fs            1.3.1       2019-05-06 [1] CRAN (R 3.6.0)                     
#>  glue          1.3.1       2019-03-12 [1] CRAN (R 3.6.0)                     
#>  highr         0.8         2019-03-20 [1] CRAN (R 3.6.0)                     
#>  htmltools     0.4.0       2019-10-04 [1] CRAN (R 3.6.0)                     
#>  ITKR          0.5.3       2020-03-18 [1] Github (stnava/ITKR@715331d)       
#>  knitr         1.28        2020-02-06 [1] CRAN (R 3.6.0)                     
#>  lattice       0.20-38     2018-11-04 [1] CRAN (R 3.6.3)                     
#>  magrittr      1.5         2014-11-22 [1] CRAN (R 3.6.0)                     
#>  Matrix        1.2-18      2019-11-27 [1] CRAN (R 3.6.3)                     
#>  pillar        1.4.3       2019-12-20 [1] CRAN (R 3.6.0)                     
#>  pkgconfig     2.0.3       2019-09-22 [1] CRAN (R 3.6.0)                     
#>  purrr         0.3.3       2019-10-18 [1] CRAN (R 3.6.0)                     
#>  Rcpp          1.0.4.3     2020-03-20 [1] Github (RcppCore/Rcpp@ce5538b)     
#>  RcppEigen     0.3.3.7.0   2019-11-16 [1] CRAN (R 3.6.0)                     
#>  reprex        0.3.0.9001  2020-01-05 [1] Github (tidyverse/reprex@5ae0b29)  
#>  rlang         0.4.5       2020-03-01 [1] CRAN (R 3.6.2)                     
#>  rmarkdown     2.1         2020-01-20 [1] CRAN (R 3.6.0)                     
#>  rstudioapi    0.11.0-9000 2020-02-19 [1] Github (rstudio/rstudioapi@deb9c47)
#>  sessioninfo   1.1.1       2018-11-05 [1] CRAN (R 3.6.0)                     
#>  stringi       1.4.6       2020-02-17 [1] CRAN (R 3.6.0)                     
#>  stringr       1.4.0       2019-02-10 [1] CRAN (R 3.6.0)                     
#>  styler        1.3.2       2020-02-23 [1] CRAN (R 3.6.2)                     
#>  tibble        2.1.3       2019-06-06 [1] CRAN (R 3.6.0)                     
#>  withr         2.1.2       2018-03-15 [1] CRAN (R 3.6.0)                     
#>  xfun          0.12        2020-01-13 [1] CRAN (R 3.6.0)                     
#>  yaml          2.2.1       2020-02-01 [1] CRAN (R 3.6.0)                     
#> 
#> [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

</details>

# Now with > 0

``` r
library(ANTsRCore)
#> 
#> Attaching package: 'ANTsRCore'
#> The following objects are masked from 'package:stats':
#> 
#>     sd, var
#> The following objects are masked from 'package:base':
#> 
#>     all, any, apply, max, min, prod, range, sum
fname = getANTsRData("r16")
in_img = antsImageRead(fname)
n4 = n4BiasFieldCorrection(in_img)
rm(n4); gc()
#>           used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells 1849052 98.8    3535897 188.9         NA  2689078 143.7
#> Vcells 3302303 25.2    8388608  64.0      16384  5643593  43.1
mask = in_img > 0
mask
#> antsImage
#>   Pixel Type          : unsigned char 
#>   Components Per Pixel: 1 
#>   Dimensions          : 256x256 
#>   Voxel Spacing       : 1x1 
#>   Origin              : 0 0 
#>   Direction           : 1 0 0 1
sum(mask)
#> [1] 19278
```

<sup>Created on 2020-03-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

<details style="margin-bottom:10px;">

<summary>Session info</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.3 (2020-02-29)
#>  os       macOS Mojave 10.14.6        
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2020-03-20                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date       lib source                             
#>  ANTsRCore   * 0.7.3.1     2020-03-20 [1] Github (ANTsX/ANTsRCore@c53aaaa)   
#>  assertthat    0.2.1       2019-03-21 [1] CRAN (R 3.6.0)                     
#>  backports     1.1.5       2019-10-02 [1] CRAN (R 3.6.0)                     
#>  cli           2.0.2       2020-02-28 [1] CRAN (R 3.6.0)                     
#>  crayon        1.3.4       2017-09-16 [1] CRAN (R 3.6.0)                     
#>  digest        0.6.25      2020-02-23 [1] CRAN (R 3.6.0)                     
#>  evaluate      0.14        2019-05-28 [1] CRAN (R 3.6.0)                     
#>  fansi         0.4.1       2020-01-08 [1] CRAN (R 3.6.0)                     
#>  fs            1.3.1       2019-05-06 [1] CRAN (R 3.6.0)                     
#>  glue          1.3.1       2019-03-12 [1] CRAN (R 3.6.0)                     
#>  highr         0.8         2019-03-20 [1] CRAN (R 3.6.0)                     
#>  htmltools     0.4.0       2019-10-04 [1] CRAN (R 3.6.0)                     
#>  ITKR          0.5.3       2020-03-18 [1] Github (stnava/ITKR@715331d)       
#>  knitr         1.28        2020-02-06 [1] CRAN (R 3.6.0)                     
#>  lattice       0.20-38     2018-11-04 [1] CRAN (R 3.6.3)                     
#>  magrittr      1.5         2014-11-22 [1] CRAN (R 3.6.0)                     
#>  Matrix        1.2-18      2019-11-27 [1] CRAN (R 3.6.3)                     
#>  pillar        1.4.3       2019-12-20 [1] CRAN (R 3.6.0)                     
#>  pkgconfig     2.0.3       2019-09-22 [1] CRAN (R 3.6.0)                     
#>  purrr         0.3.3       2019-10-18 [1] CRAN (R 3.6.0)                     
#>  Rcpp          1.0.4.3     2020-03-20 [1] Github (RcppCore/Rcpp@ce5538b)     
#>  RcppEigen     0.3.3.7.0   2019-11-16 [1] CRAN (R 3.6.0)                     
#>  reprex        0.3.0.9001  2020-01-05 [1] Github (tidyverse/reprex@5ae0b29)  
#>  rlang         0.4.5       2020-03-01 [1] CRAN (R 3.6.2)                     
#>  rmarkdown     2.1         2020-01-20 [1] CRAN (R 3.6.0)                     
#>  rstudioapi    0.11.0-9000 2020-02-19 [1] Github (rstudio/rstudioapi@deb9c47)
#>  sessioninfo   1.1.1       2018-11-05 [1] CRAN (R 3.6.0)                     
#>  stringi       1.4.6       2020-02-17 [1] CRAN (R 3.6.0)                     
#>  stringr       1.4.0       2019-02-10 [1] CRAN (R 3.6.0)                     
#>  styler        1.3.2       2020-02-23 [1] CRAN (R 3.6.2)                     
#>  tibble        2.1.3       2019-06-06 [1] CRAN (R 3.6.0)                     
#>  withr         2.1.2       2018-03-15 [1] CRAN (R 3.6.0)                     
#>  xfun          0.12        2020-01-13 [1] CRAN (R 3.6.0)                     
#>  yaml          2.2.1       2020-02-01 [1] CRAN (R 3.6.0)                     
#> 
#> [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

</details>

**Comment by ntustison** [comment]
So none of the normal screen dump gets printed to the screen?

**Comment by muschellij2** [comment]
It's hard to demonstrate, but I made a quick video: https://youtu.be/dVcidFCeTZ0

**Comment by ntustison** [comment]
Awesome.   Thanks for the video.  I'm pretty sure I found it although it's not obvious.  Let me push to the ANTs repository, I'll point out the issue, you can update the ANTs git tag, and see if it works for you.

**Comment by ntustison** [comment]
Here you go.  I'm guessing [this](https://github.com/ANTsX/ANTs/commit/0a4fedf3375ce8fd534b0df57abd24095feeac8a#diff-06b7de4cbfdd1e2cd545c132a73b8a4eL128-R128) is the culprit.

**Comment by muschellij2** [comment]
Installing now - will let you know if same behavior develops.

**Pull request closed without merging**